File: 자바스크립트/예산(연습문제).js
function solution(budgets, M) {
    let left = 1;
    let right = Math.max(...budgets);
    while (left < right) {
        const middle = parseInt((left + right + 1) / 2)
        const required = budgets.map(budget => Math.min(middle, budget))
        const sum = required.reduce((total, num) => total + num, 0);
        sum <= M ? left = middle : right = middle - 1
    }
    return left;
}

File: 파이썬/예산(연습문제).py
# 이분탐색 - 최대값
def is_poss(middle, budgets, M):
    return M >= sum([min(middle, budget) for budget in budgets])


def solution(budgets, M):
    left, right = 1, max(budgets)
    while left < right:
        middle = (left + right + 1) // 2
        if is_poss(middle, budgets, M):
            left = middle
        else:
            right = middle - 1
    return right


# 정렬
'''
def solution(budgets, M):
    max_require = max(budgets) 
    budgets.sort(reverse=True)
    
    while budgets:
        if budgets[-1] * len(budgets) < M:
            M -= budgets.pop()
        else:
            return M // len(budgets)
    return max_require
'''
